### PR TITLE
plugin(grpc-transcode): support new options for `protocol buffer` encoding.

### LINKED ISSUE #821: Bug: the problem of data precision loss when grpc transcode has Int64 type in returned data
### Environment

* apisix version : 0.8
* OS: CentOS 7.6

最近试用zeebe, 因其Gateway协议为gRPC, 所以用grpc-transcode 做一个测试, 发现回传的int64类型的数据精度有损失, 测试结果如下:

```
curl -i "http://127.0.0.1:9080/zeebe/WorkflowInstanceCreate?bpmnProcessId=order-process&version=1&variables=\{\"orderId\":\"7\",\"ordervalue\":99\}"
HTTP/1.1 200 OK
Date: Wed, 06 Nov 2019 10:43:35 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
grpc-encoding: identity
grpc-accept-encoding: gzip
Server: APISIX web server
Trailer: grpc-status
Trailer: grpc-message

{"workflowKey":2.2517998136853e+15,"workflowInstanceKey":2.2517998136879e+15,"bpmnProcessId":"order-process","version":1}
```
之后查阅源码, 发现是在response中调用 lua-protobuf 的 pb.decode 造成的损失; 同时发现在lua-protobuf中有一个option的函数可以在decode 或是 encode之前调用将int64 转换成 string, 来缓解精度损失的问题;
其参数定义如下:

```lua
static int Lpb_option(lua_State *L) {
#define OPTS(X) \
    X(0, enum_as_name,          LS->enum_as_value = 0)             \
    X(1, enum_as_value,         LS->enum_as_value = 1)             \
    X(2, int64_as_number,       LS->int64_mode = LPB_NUMBER)       \
    X(3, int64_as_string,       LS->int64_mode = LPB_STRING)       \
    X(4, int64_as_hexstring,    LS->int64_mode = LPB_HEXSTRING)    \
    X(5, auto_default_values,   LS->default_mode = LPB_DEFDEF)     \
    X(6, no_default_values,     LS->default_mode = LPB_NODEF)      \
    X(7, use_default_values,    LS->default_mode = LPB_COPYDEF)    \
    X(8, use_default_metatable, LS->default_mode = LPB_METADEF)    \
    X(9, enable_hooks,          LS->use_hooks = 1)                 \
    X(10, disable_hooks,        LS->use_hooks = 0)                 \
```

所以临时在response.lua 的 pb.decode 之前加一个 pb.option("int64_as_string"), 测试了一下, 结果如下:

```
 curl -i "http://127.0.0.1:9080/zeebe/WorkflowInstanceCreate?bpmnProcessId=order-process&version=1&variables=\{\"orderId\":\"9\",\"ordervalue\":99\}"
HTTP/1.1 200 OK
Date: Fri, 08 Nov 2019 03:28:06 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
grpc-encoding: identity
grpc-accept-encoding: gzip
Server: APISIX web server
Trailer: grpc-status
Trailer: grpc-message

{"workflowKey":"#2251799813685260","workflowInstanceKey":"#2251799813687941","bpmnProcessId":"order-process","version":1}
```

看起来精度损失的问题有解了.

现在我想在grpc-transcode的conf 中增加一个配置项, 来解决未来其它使用者的此类问题, 还需大家提供点建议!

### PR DESCRIPTION
**Pull request by tom2nonames**
### Summary

主要是处理了int64类型数据精度损失的问题

### Issues resolved

Fix #821

### CODE REVIEW

**Comment by nic-chen** [comment]
@tom2nonames  need test cases cover.

**Comment by tom2nonames** [comment]
Dear nic-chen,


测试用例本打算写的, 但这个测试需要有一个支持grpc的服务, 需要定义proto的message 和 rpc 方法, 我在源代码中也没有找到这一部分的代码, 我想CI 跑的时候这个服务需要先启动吧! 


所以不知道从何处下手.


刚刚找了一圈, 看到了这个https://github.com/nic-chen/grpc_server_example


我再看看, 如果是基于这个服务的话,我去试着写一下. 


tom2nonames
2019-11-


 原始邮件 
发件人: nic-chen<notifications@github.com>
收件人: apache/incubator-apisix<incubator-apisix@noreply.github.com>
抄送: tom2nonames<13682645264@139.com>; Mention<mention@noreply.github.com>
发送时间: 2019年11月13日(周三) 19:06
主题: Re: [apache/incubator-apisix] grpc-transcode 插件的配置中增加proto option 选项开关. (#846)


@tom2nonames need test cases cover.

File: doc/plugins/grpc-transcoding-cn.md [hunk 1]
```diff
@@ -107,3 +107,72 @@ Proxy-Connection: keep-alive
...
+{
+    "methods": ["GET"],
+    "uri": "/zeebe/WorkflowInstanceCreate",
+    "service_protocol": "grpc",
+    "plugins": {
+        "grpc-transcode": {
+         "proto_id": "1",
+         "service": "gateway_protocol.Gateway",
+         "method": "CreateWorkflowInstance",
+         "option":["int64_as_string"]
```
Comment (membphis) [inline doc/plugins/grpc-transcoding-cn.md]: Missing schema to check it.

**Comment by nic-chen** [comment]
@tom2nonames 

grpc-transcode的测试用例，你可以在里面添加：
 We have test cases for grpc-transcode , you can add some here:

https://github.com/apache/incubator-apisix/blob/master/t/plugin/grpc-transcode.t

File: doc/plugins/grpc-transcoding-cn.md (hunk 1 above)
Comment (membphis) [reply doc/plugins/grpc-transcoding-cn.md]: `pb_options` is a better name.

Comment (tom2nonames) [reply doc/plugins/grpc-transcoding-cn.md]: Schema check increased, variable name replaced with "pb_option", Test case is trying.

File: lua/apisix/plugins/grpc-transcode.lua [hunk 2]
```diff
@@ -16,18 +16,53 @@
...
 local proto       = require("apisix.plugins.grpc-transcode.proto")
 local request     = require("apisix.plugins.grpc-transcode.request")
 local response    = require("apisix.plugins.grpc-transcode.response")
 
 
+local pb_option_def = {
+    {   description = "enum as result",
+        type = "string",
+        enum = {"int64_as_number",
+              "int64_as_string",
```
Comment (membphis) [inline lua/apisix/plugins/grpc-transcode.lua]: Code style: bad indentation.

File: lua/apisix/plugins/grpc-transcode.lua [hunk 3]
```diff
@@ -16,18 +16,53 @@
...
 local core        = require("apisix.core")
+local schema_def  = require("apisix.schema_def")
 local plugin_name = "grpc-transcode"
 local proto       = require("apisix.plugins.grpc-transcode.proto")
 local request     = require("apisix.plugins.grpc-transcode.request")
 local response    = require("apisix.plugins.grpc-transcode.response")
 
 
+local pb_option_def = {
+    {   description = "enum as result",
```
Comment (membphis) [inline lua/apisix/plugins/grpc-transcode.lua]: Code style

`description` should be placed on a new line.

**Comment by membphis** [comment]
@nic-chen what is your opinion?

File: doc/plugins/grpc-transcoding-cn.md [hunk 4]
```diff
@@ -107,3 +107,72 @@ Proxy-Connection: keep-alive
...
+    "upstream": {
+        "type": "roundrobin",
+        "nodes": {
+            "127.0.0.1:26500": 1
+        }
+    }
+}'
+```
+
+#### 测试
```
Comment (moonming) [inline doc/plugins/grpc-transcoding-cn.md]: need add test cases in Travis CI.

**Comment by nic-chen** [comment]
> @nic-chen what is your opinion?

I think it's almost ok after appending test cases..

**Comment by moonming** [comment]
@tom2nonames any update?

**Comment by moonming** [comment]
@nic-chen can you add test cases for this PR? so we can merge it first.

**Pull request merged**